src/components/SettingsBox/SettingsBox.js

```diff
@@ -20,49 +20,65 @@ export default class SettingsBox extends Component {
     }
 
     render() {
-        const headers = this.props.headers;
+        const view = {
```

Comment:
I reckoned the content (from RequestBuilder) should be separate from how the content is rendered (in SettingsBox). I think having the display information in the RequestBuilder clogs up RequestBuilder, as it mainly defines callbacks to be passed into its children components which do the rendering of those props. Originally, the headers were split across data.js and RequestBuilder, which looked redundant. 

I was inspired by some semblance of Model-View-Controller, so I wanted to have at least the view separated into the child. The model (?) and the controller is in the parent. 

One way I can think of the headers being in one place (`display`, `type`, and `value` keys) is by wrapping the original `value`s in lambdas, like so `() =>`. I suppose I'd still end up having to pass the state in somehow. Maybe it's cleaner than mapping?

Reply:
SettingsBox is a stateless react component, it only renders.  All information it needs to render should be passed to it through props, not maintained inside the component.  

I don't see why the display, type, and value can't all just be in `RequestBuilder`.  In fact the whole thing should be part of the state which is both propagated to dependents and modified within the state.  

That is to say, there should be no redundancy between the state of the RequestBuilder, where it first defines something e.g. `alternativeTherapy`, and the rendering information e.g. type/display/value.  It should all just be part of the state or defined in one place, in a list or object, with no separation at all.

The only issue would be "settings" which are buttons, but those should not be with the rest of the settings anyway.  They don't modify the state, they just use it, and they break the pattern of callback to state update which is confusing.  They should be pulled out and put into the settings box along with their associated functions, or they could be separated out into another prop that gets passed to SettingsBox to be rendered, separate from the actual settings.

Reply:
I agree with this part of your explanation especially (not to say I do not agree with the text before it, just drawing out this part):
> It should all just be ... defined in one place, in a list or object, with no separation at all. 

I don't think the `display` and `type` should be in the state since there's never a way for the user to modify the input/checkbox headers or how the content should be rendered (as an input or a checkbox). But I think all the React.js components (like SettingsBox) should fall under the category of a "controller" and the raw, changeable strings that are the `display` could be moved to some place like data.js. React.js sort of blurs the lines of MVC anyways... 

> The only issue would be "settings" which are buttons, but those should not be with the rest of the settings anyway. They don't modify the state, they just use it, and they break the pattern of callback to state update which is confusing. They should be pulled out and put into the settings box along with their associated functions, or they could be separated out into another prop that gets passed to SettingsBox to be rendered, separate from the actual settings.
I agree. I've not yet brainstormed an approach to this- to me the reset buttons are part of the "kitchen utilities/junk drawer" of the SettingsBox component. @plarocque4 what are your thoughts?

Reply:
sure, don't put the display and the type in the state.  It's not particularly important.  If the settings are to be derived from some JSON object, then the state should be derived from the same object.  There should be no ambiguity between the state and the definition about what the settings are named.  

If you have a JSON definition of name/type/initialValue/display, you can generate the state of the settingsBox from that JSON, using the name and initialValue, and then pass the JSON itself into settingsBox.  Now the state and the settings are synced.  

Right now we are halfway between a dynamic and static approach.

Reply:
sure, don't put the display and the type in the state. It's not particularly important. If the settings are to be derived from some JSON object, then the state should be derived from the same object. There should be no ambiguity between the state and the definition about what the settings are named.

If you have a JSON definition of name/type/initialValue/display, you can generate the state of the settingsBox from that JSON, using the name and initialValue, and then pass the JSON itself into settingsBox. Now the state and the settings are synced.

Right now we are halfway between a dynamic and static approach.

Reply:
I... missed this comment, sorry! 

Your comment reminds me of the latest React API (with the functional components) and passing an initial state in and using a reducer. Maybe I can try doing that in that request-generator refactoring ticket one day.